NIFI-10010: ListenTCP adds client certificate's Subject and Issuer DN…

pgyori opened the pull request:
…s to flowfiles as attributes

<!-- Licensed to the Apache Software Foundation (ASF) under one or more -->
<!-- contributor license agreements.  See the NOTICE file distributed with -->
<!-- this work for additional information regarding copyright ownership. -->
<!-- The ASF licenses this file to You under the Apache License, Version 2.0 -->
<!-- (the "License"); you may not use this file except in compliance with -->
<!-- the License.  You may obtain a copy of the License at -->
<!--     http://www.apache.org/licenses/LICENSE-2.0 -->
<!-- Unless required by applicable law or agreed to in writing, software -->
<!-- distributed under the License is distributed on an "AS IS" BASIS, -->
<!-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. -->
<!-- See the License for the specific language governing permissions and -->
<!-- limitations under the License. -->

# Summary

https://issues.apache.org/jira/browse/NIFI-10010
Implements NIFI-10010. ListenTCP adds the client certificate's Subject DN and Issuer DN values to the outgoing flowfiles as attributes in case of secure connections.

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [ ] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 8
  - [ ] JDK 11
  - [ ] JDK 17

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [x] Documentation formatting appears as expected in rendered files

exceptionfactory commented on nifi-nar-bundles/nifi-extension-utils/nifi-event-transport/src/main/java/org/apache/nifi/event/transport/SslSessionStatus.java:
```diff
@@ -0,0 +1,36 @@
+package org.apache.nifi.event.transport;
+
+public class SslSessionStatus {
+    private final String subjectDN;
+    private final String issuerDN;
```
Recommend using `X500Principal` instead of `String` and renaming these properties to just `subject` and `issuer` in order to provide better fidelity with the `X509Certificate` source of the information.

exceptionfactory commented on nifi-nar-bundles/nifi-extension-utils/nifi-event-transport/src/main/java/org/apache/nifi/event/transport/netty/codec/SocketByteArrayMessageDecoder.java:
```diff
@@ -38,7 +47,38 @@ public class SocketByteArrayMessageDecoder extends MessageToMessageDecoder<byte[
+            final Certificate[] certificates = sslSession.getPeerCertificates();
+            if (certificates.length > 0) {
+                final X509Certificate certificate = (X509Certificate) certificates[0];
+                final String subjectDN = certificate.getSubjectDN().toString();
+                final String issuerDN = certificate.getIssuerDN().toString();
```
Following the comment to change the `SslSessionStatus` property types, recommend replacing `getSubjectDN()` and `getIssuerDN()` with `getSubjectX500Principal()` and `getIssuerX500Principal()` since the DN methods are denigrated according to the Java 8 documentation:

https://docs.oracle.com/javase/8/docs/api/java/security/cert/X509Certificate.html#getIssuerDN--

exceptionfactory commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ListenTCP.java:
```diff
@@ -291,4 +305,15 @@ protected String getBatchKey(ByteArrayMessage event) {
+        if (sslSessionStatus != null) {
+            attributes.put(CLIENT_CERTIFICATE_SUBJECT_DN_ATTRIBUTE, sslSessionStatus.getSubjectDN());
+            attributes.put(CLIENT_CERTIFICATE_ISSUER_DN_ATTRIBUTE, sslSessionStatus.getIssuerDN());
+        } else {
+            getLogger().debug("Remote Peer [{}] not verified: client certificates not provided",
+                    event.getSender());
+        }
```
Having the debug log seems unnecessary since the absence of the attributes in outgoing FlowFiles indicates the same fact of missing client certificate information.
```suggestion
        }
```

exceptionfactory reviewed (changes requested):
Thanks for putting together this improvement @pgyori! The general approach looks good, just noted a few naming and implementation adjustments.

pgyori commented on nifi-nar-bundles/nifi-extension-utils/nifi-event-transport/src/main/java/org/apache/nifi/event/transport/netty/codec/SocketByteArrayMessageDecoder.java:
```diff
@@ -38,7 +47,38 @@ public class SocketByteArrayMessageDecoder extends MessageToMessageDecoder<byte[
+        ByteArrayMessage message = getMessageWithSslSessionStatus(channelHandlerContext, bytes, address);
+        if (message == null) {
+            message = new ByteArrayMessage(bytes, address);
+        }
```
This is a great idea, makes it much cleaner. Thanks!

exceptionfactory commented on nifi-nar-bundles/nifi-extension-utils/nifi-event-transport/src/main/java/org/apache/nifi/event/transport/netty/codec/SocketByteArrayMessageDecoder.java:
```diff
@@ -38,7 +48,37 @@ public class SocketByteArrayMessageDecoder extends MessageToMessageDecoder<byte[
     }
+
+    private SslSessionStatus getSslSessionStatus(final ChannelHandlerContext channelHandlerContext) {
+        final Iterator<Map.Entry<String, ChannelHandler>> iterator = channelHandlerContext.channel().pipeline().iterator();
+        while (iterator.hasNext()) {
+            final ChannelHandler channelHandler = iterator.next().getValue();
+            if (channelHandler instanceof SslHandler) {
+                return createSslSessionStatusFromSslHandler((SslHandler) channelHandler);
+            }
+        }
+        return null;
```
This could be refactored to use a single return statement. ChannelPipeline also implements `Iterable`, so it could be streamlined.
```suggestion
        SslHandler sslHandler = null;
        for (final Map.Entry<String, ChannelHandler> entry : channelHandlerContext.channel().pipeline()) {
            final ChannelHandler channelHandler = entry.getValue();
            if (channelHandler instanceof SslHandler) {
                sslHandler = (SslHandler) channelHandler;
                break;
            }
        }
        return sslHandler == null ? createSslSessionStatusFromSslHandler(sslHandler);
```

pgyori replied:
Thank you! I applied your suggested change.

exceptionfactory reviewed (changes requested):
Thanks for making the updates @pgyori! This looks close the completion. I noted two recommended improvements.

Other than those changes, can you update `TestListenTCP` to include checks for the new attributes? It should be possible to update `testRunClientAuthRequired()` and check for the existence of the new attributes.

exceptionfactory commented on nifi-nar-bundles/nifi-extension-utils/nifi-event-transport/pom.xml:
```diff
@@ -38,5 +38,9 @@
             <version>1.17.0-SNAPSHOT</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+        </dependency>
```
This JUnit 4 dependency should be removed.

pgyori commented:
Thank you again, @exceptionfactory .
I extended the test case in TestListenTCP as you recommended and also added a check for the new attributes not being present when the connection is not secure.
I added a test class for SocketByteArrayMessageDecoder with test cases.
Applied your recommendations almost exactly as you requested. Please check my modifications. Thanks!

exceptionfactory commented on nifi-nar-bundles/nifi-extension-utils/nifi-event-transport/src/test/java/org/apache/nifi/event/transport/netty/SocketByteArrayMessageDecoderTest.java:
```diff
@@ -0,0 +1,166 @@
+        initNoSslHandler();
+
+        List<Object> decoded = new ArrayList<>(1);
+        decode(channelHandlerContext, "test".getBytes(), decoded);
```
The `"test".getBytes()` method should be declared statically, and `getBytes()` should include the Standard UTF-8 character set.

exceptionfactory commented on nifi-nar-bundles/nifi-extension-utils/nifi-event-transport/src/test/java/org/apache/nifi/event/transport/netty/SocketByteArrayMessageDecoderTest.java:
```diff
@@ -0,0 +1,166 @@
+
+    private void initSsl() throws Exception {
+        initSslEngine();
+        try (InputStream inStream = new FileInputStream("src/test/resources/netty/client_certificate.cer")) {
+            CertificateFactory cf = CertificateFactory.getInstance("X.509");
+            certificate = (X509Certificate)cf.generateCertificate(inStream);
+        }
```
Static certificate files should not be checked into version control as they will eventually expire and lead to test failures. See `TemporaryKeyStoreBuilder` for an approach to generating certificates for testing.

exceptionfactory commented on nifi-nar-bundles/nifi-extension-utils/nifi-event-transport/src/test/resources/netty/client_certificate.cer:
```diff
@@ -0,0 +1,18 @@
+-----BEGIN CERTIFICATE-----
+MIIC3jCCAcagAwIBAgIEYn5khzANBgkqhkiG9w0BAQsFADAvMRwwGgYDVQQKDBNJ
+c3N1ZXIgT3JnYW5pemF0aW9uMQ8wDQYDVQQDDAZJc3N1ZXIwIBcNMjIwNTEzMTQw
+MDM5WhgPMzAyMTA1MTMxNDAwMzlaMDExHTAbBgNVBAoMFFN1YmplY3QgT3JnYW5p
+emF0aW9uMRAwDgYDVQQDDAdTdWJqZWN0MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8A
+MIIBCgKCAQEA5I2jZf5rr6nCckKfO/9EZnWkD5AjB2P0P1HJQ3OrjHAmfzZGe6mK
+M2pwn+k/dqswkg6lU2F+Ey+SGnK9Y5AJd35nBN3p+jV+BGsxAjxkhdVgOlBK0GyB
+E5n9dcXiu9DrFgif6PwPjQe4eshXRpASTrI/ZPaIXf6m8Jy6gahThNEFGmDwRyll
+iN/29fglii3v+Ih1eRes2gH6LyRRS1jnVVlbZUvOXK+gqY+YqpAhOAIsdNb3zzJ/
+hGcuIfh4yWQwii9xWzFhinOstHFIOCvdf+DlU5LPS55XYq1rZLzTucwot+f5nDYJ
+rvUWzUuqQVGboQfeyuwkbMyoy8SaIeyCpwIDAQABMA0GCSqGSIb3DQEBCwUAA4IB
+AQA3Qw+00q7+NKy3jzz+CTDyFehMZvDo8cDO0xGm2qB4hjcR2Sn0oHCbKMZGjwOv
+U3Aprqqn9/9+hEVtHj7HmcBM4ocbqAXhdKhKOslPYTt9NRLLyHxrdqPQwQ4AzLGG
+GW4kpQQZa1GCaOyXOuAz+LV+vAuFIyvq7PH98yaCIxXrJ2QxhceGZnTtYKSwRnhj
+uoG3R7dwumXjiXI7JzhgxX2WQH6qSx1it/PsI4JqghK1ML0xtH/vH+pY9FbODHta
+K5FnWnS6B35Qbrei1Irf7LmzZU9C+feZOe0MvtSBrbpcX3BSIHBJgwM179MtwYZM
+sFxJtzcHUGnl0K0fwhrs12h+
+-----END CERTIFICATE-----
```
As mentioned, this file should be removed.

pgyori commented on nifi-nar-bundles/nifi-extension-utils/nifi-event-transport/src/main/java/org/apache/nifi/event/transport/netty/codec/SocketByteArrayMessageDecoder.java:
```diff
@@ -38,7 +48,37 @@ public class SocketByteArrayMessageDecoder extends MessageToMessageDecoder<byte[
+        } catch (SSLPeerUnverifiedException peerUnverifiedException) {
+            return null;
+        }
+        return null;
```
Unfortunately there is no access to a logger in SocketByteArrayMessageDecoder.
I refactored the method to have only one return statement, because I agree that it looks cleaner. The catch clause is now empty, which I know is not nice, but I'm not sure we can do anything in that block.

exceptionfactory commented on nifi-nar-bundles/nifi-extension-utils/nifi-event-transport/src/test/java/org/apache/nifi/event/transport/netty/SocketByteArrayMessageDecoderTest.java:
```diff
@@ -0,0 +1,166 @@
+
+    private InetSocketAddress inetSocketAddress;
+    private X509Certificate certificate;
+    @Mock
+    private ChannelHandlerContext channelHandlerContext;
+    @Mock
+    private Channel channel;
+    @Mock
+    private ChannelPipeline channelPipeline;
+    @Mock
+    private Iterator<Map.Entry<String, ChannelHandler>> iterator;
+    @Mock
+    private Map.Entry<String, ChannelHandler> pipelineEntry;
+    @Mock
+    private SslHandler sslHandler;
+    @Mock
+    private ChannelOutboundHandlerAdapter nonSslHandler;
+    @Mock
+    private SSLEngine sslEngine;
+    @Mock
+    private SSLSession sslSession;
```
Although this test is narrowly scoped, the number of mocked components makes it difficult to maintain. It would be better to test this class on conjunction with a functional Netty implementation. With the usage of this class tested through TestListenTCP, it seems better to remove this unit test.

exceptionfactory reviewed (changes requested):
Thanks for the additional updates @pgyori. Unfortunately the new test for `SocketByteArrayMessageDecoder` seems difficult to maintain for several reasons. The static certificate file is a problem for future maintenance, and the number of mocked components in the unit test is also an issue. Although having a unit test at this level is helpful, it would be better to test the decoder in the context of a running socket server, given the behavior of Netty itself. For these reasons, I don't think a unit test for the decoder is necessary, and would prefer to merge the other changes without the unit test class.

pgyori commented:
I understand your concerns about the test for SocketByteArrayMessageDecoder. I'm OK with omitting that commit. Thanks for the quick feedback!

exceptionfactory reviewed (approved):
Thanks for working through the feedback @pgyori, looks good! +1 merging.
